Samples/DirectMLNpuInference/main.cpp

```diff
@@ -0,0 +1,177 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include "pch.h"
+
+#include <dxcore_interface.h>
+#include <dxcore.h>
+
+#include "onnxruntime_cxx_api.h"
+#include "dml_provider_factory.h"
+
+#include "TensorHelper.h"
+
+using Microsoft::WRL::ComPtr;
+
+void InitializeDirectML(ID3D12Device1** d3dDeviceOut, ID3D12CommandQueue** commandQueueOut, IDMLDevice** dmlDeviceOut) {
+    // Whether to skip adapters which support Graphics in order to target NPU for testing
+    bool forceComputeOnlyDevice = true;
+    ComPtr<IDXCoreAdapterFactory> factory;
+    HMODULE dxCoreModule = LoadLibraryW(L"DXCore.dll");
+    if (dxCoreModule)
+    {
+        auto dxcoreCreateAdapterFactory = reinterpret_cast<HRESULT(WINAPI*)(REFIID, void**)>(
+            GetProcAddress(dxCoreModule, "DXCoreCreateAdapterFactory")
+            );
+        if (dxcoreCreateAdapterFactory)
+        {
+            dxcoreCreateAdapterFactory(IID_PPV_ARGS(&factory));
+        }
+    }
+    // Create the DXCore Adapter
+    ComPtr<IDXCoreAdapter> adapter;
+    if (factory)
+    {
+        const GUID dxGUIDs[] = { DXCORE_ADAPTER_ATTRIBUTE_D3D12_CORE_COMPUTE };
+        ComPtr<IDXCoreAdapterList> adapterList;
+        THROW_IF_FAILED(factory->CreateAdapterList(ARRAYSIZE(dxGUIDs), dxGUIDs, IID_PPV_ARGS(&adapterList)));
+        for (uint32_t i = 0, adapterCount = adapterList->GetAdapterCount(); i < adapterCount; i++)
+        {
+            ComPtr<IDXCoreAdapter> nextGpuAdapter;
+            THROW_IF_FAILED(adapterList->GetAdapter(static_cast<uint32_t>(i), IID_PPV_ARGS(&nextGpuAdapter)));
+            if (!forceComputeOnlyDevice || !nextGpuAdapter->IsAttributeSupported(DXCORE_ADAPTER_ATTRIBUTE_D3D12_GRAPHICS))
+            {
+                adapter = std::move(nextGpuAdapter);
+                break;
+            }
+        }
+    }
+    // Create the D3D12 Device
+    ComPtr<ID3D12Device1> d3dDevice;
+    if (adapter)
+    {
+        HMODULE d3d12Module = LoadLibraryW(L"d3d12.dll");
+        if (d3d12Module)
+        {
+            auto d3d12CreateDevice = reinterpret_cast<HRESULT(WINAPI*)(IUnknown*, D3D_FEATURE_LEVEL, REFIID, void*)>(
+                GetProcAddress(d3d12Module, "D3D12CreateDevice")
+                );
+            if (d3d12CreateDevice)
+            {
+                THROW_IF_FAILED(d3d12CreateDevice(adapter.Get(), D3D_FEATURE_LEVEL_1_0_CORE, IID_PPV_ARGS(&d3dDevice)));
+            }
+        }
+    }
+    // Create the DML Device and D3D12 Command Queue
+    ComPtr<IDMLDevice> dmlDevice;
+    ComPtr<ID3D12CommandQueue> commandQueue;
+    if (d3dDevice)
+    {
+        D3D12_COMMAND_QUEUE_DESC queueDesc = {};
+        queueDesc.Type = D3D12_COMMAND_LIST_TYPE_COMPUTE;
+        THROW_IF_FAILED(d3dDevice->CreateCommandQueue(
+            &queueDesc,
+            IID_PPV_ARGS(commandQueue.ReleaseAndGetAddressOf())));
+        HMODULE dmlModule = LoadLibraryW(L"DirectML.dll");
```

Comment:
Any particular reason to LoadLibrary this? I believe the DML NuGet package should automatically add directml.lib as a static dependency.

Reply:
Taken from the snippet the Jeff shared that was provided in the dev-kit documentation last year